GDevelop Multiplayer docs (#191)
GDevelopApp/GDevelop-documentation: docs/gdevelop5/all-features/multiplayer/index.md

---
title: GDevelop Multiplayer
---

# How can I add multiplayer features to my game?

Similar to [Player Authentication](/gdevelop5/all-features/player-authentication) and Leaderboards, GDevelop provides a built-in solution to add **real-time multiplayer features** to your game, whether you're making a simple multiplayer game or a more complex one, from a cooperative game to a competitive one.

It is built to be easy to use, accessible to everyone, whilst providing a good experience for all your players. 

Some features available out of the box are:

- Lobbies for your game and a user interface allowing players to start a new game,
- Built-in authentication for players to create an account and login to your game,
- Automatic choice of the host of the game,
- Automatic synchronization of players, depending on who is in charge of each object,
- Automatic synchronization of the game state (game objects, variables...)

!!! note

> 

Reviewer: Remove this line to put the paragraph in the note

   Usually, adding multiplayer features or creating from scratch a multiplayer game is something that is difficult and error-prone. Most solutions, including some you could adapt or use with GDevelop, requires some programming and networking knowledge. With GDevelop Multiplayer features, there is **no need to handle servers**, sending custom messages, handling connections, lobbies, packets being lost, prediction of movements, interpolation, etc.
   GDevelop Multiplayer infrastructure is built to scale from very small game to games with thousands of concurrent players. If you're a game studio or an agency with specific needs (for example, launching a game for a promotional campaign to an audience of dozens of thousands of people), feel free to reach out to discuss your specific use-case.

# How to use

> To use the multiplayer features, access the behaviors and actions/conditions of the **"Multiplayer"** extension in GDevelop. It is automatically available in your game, and you can start using it right away.

Reviewer: ```suggestion
To use the multiplayer features with their actions/conditions/expressions, add the behavior **"Multiplayer"** to an object in GDevelop. This behavior is automatically available in all GDevelop project, and you can start making multiplayer game right away.
```

Similar to [Player Authentication](/gdevelop5/all-features/player-authentication), GDevelop provides a built-in solution to add multiplayer features to your game, wether you're trying to create a simple multiplayer game or a more complex one, from a cooperative game to a competitive one.

It is built to be easy to use, accessible to everyone, whilst providing a good experience for all your players. (No need to worry about servers, sending custom messages, handling connections, lobbies, packets being lost, prediction of movements, interpolation, etc.)

What does it provide?

- Lobbies for your game
- Automatic authentication so you can identify players
- Automatic definition of the host of the game
- Objects are automatically synchronized between players, depending on who is in charge of each object

# How do I use it?

To use the multiplayer features, you will need to use the built-in extension "Multiplayer" in GDevelop. It is automatically available in your game, and you can start using it right away.

> It provides actions, conditions, expressions and a behavior to handle the multiplayer features in your game.

Reviewer: > built-in extension "Multiplayer" 
This is a technical detail people don't mind, for them an extension is what is in the project manager. 
I think this Multiplayer can simply be called a new feature.

Reviewer: > It is automatically available in your game, and you can start using it right away.

This is a feature bundled with the app.
To using it a multiplayer behavior must be added on objects that need to be sync.

## Lobby system

The multiplayer extension provides a lobby system, so players can join a game together. It is a key feature to allow players to play together, define which player number they are, and start the game when all players are ready.

> Lobbies are automatically created for your game (In the future, we will provide a way to customize the lobby system).

Reviewer: Maybe a screenshot here to illustrate what looks like the lobby?

- Action `Send a custom message to other players`

GDevelop Multiplayer extension automatically synchronizes objects between players, as well as variables and object information. But if you need to send a custom message to other players, for some specific cases, you can use this action.

- Condition `Custom message has been received from another player`

When a custom message has been sent, this condition will turn true for both the sender and the players who received the message, so you can handle the message the same way for everyone in your events and do something in your game.

## Synchronizing objects

The Multiplayer extension provides a behavior to synchronize objects between players.
If this behavior is not used, no objects are synchronized between players when the game starts. It means that every player will see their own version of the game.

To synchronize an object between players, you need to add the behavior `Multiplayer object` to the object.
For each synchronized object, there is a notion of "owner" of the object. The owner is the player who is in charge of the object, and who can move it, change its properties, etc.

> By default, when you use this behavior, the owner of the object is the server. But you can change the owner of an object by using the action `Player object ownership`.

Reviewer: You mention the **server**, but you didn't explain before who or how the server is defined.

Reviewer 2: Will fix this, thanks!

!!! note

    Even if player 1 is the server, there is a difference between an object owned by the server and an object owned by player 1.
    You can for instance allow another player to grab an object if the server owns it (as in "no-one owns it"), but not if player 1 owns it.

You can also remove the ownership of an object by using the action `Remove object ownership`.
When used, the object will be owned back by the server, which will start synchronizing it again between players.

### What to synchronize?

> Typically, you will want to use the behavior `Multiplayer object` on not too many objects, to avoid overloading the network and the game. You can use it on objects that are important for the game, and you want to be sure that everyone sees them the same way. For instance:

Reviewer: ```suggestion
To avoid overloading the network and the game, use the Multiplayer behavior on objects that are important to the game, or that need to be seen by everyone in the same way. For example:
```

Reviewer 2: Rewritten 👍
